File: microservices/user-service/src/middleware/logging.middleware.ts
import { Request, Response, NextFunction } from 'express';
import { getLogger } from '../utils/logger.js';

/**
 * Middleware to log all incoming requests
 */
export const loggingMiddleware = async (req: Request, res: Response, next: NextFunction) => {
  const logger = getLogger();
  const url = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
  const correlationId = req.correlationId || 'no-correlation-id';
  
  // Log the incoming request
  await logger.info(
    url,
    correlationId,
    `${req.method} ${req.originalUrl}`,
    {
      method: req.method,
      ip: req.ip,
      userAgent: req.get('user-agent')
    }
  );
  
  // Capture response
  const originalSend = res.send;
  res.send = function (data): Response {
    // Log response
    logger.info(
      url,
      correlationId,
      `Response ${res.statusCode} for ${req.method} ${req.originalUrl}`
    ).catch(console.error);
    
    return originalSend.call(this, data);
  };
  
  next();
};

File: microservices/subscription-service/Middleware/LoggingMiddleware.cs
using SubscriptionService.Logging;

namespace SubscriptionService.Middleware
{
    public class LoggingMiddleware
    {
        private readonly RequestDelegate _next;
        private readonly RabbitMQLogger _logger;

        public LoggingMiddleware(RequestDelegate next, RabbitMQLogger logger)
        {
            _next = next;
            _logger = logger;
        }

        public async Task InvokeAsync(HttpContext context)
        {
            var url = $"{context.Request.Scheme}://{context.Request.Host}{context.Request.Path}{context.Request.QueryString}";
            var correlationId = context.Items["CorrelationId"]?.ToString() ?? "no-correlation-id";

            // Log incoming request
            _logger.Info(
                url,
                correlationId,
                $"{context.Request.Method} {context.Request.Path}",
                new
                {
                    method = context.Request.Method,
                    ip = context.Connection.RemoteIpAddress?.ToString(),
                    userAgent = context.Request.Headers["User-Agent"].ToString()
                }
            );

            await _next(context);

            // Log response
            _logger.Info(
                url,
                correlationId,
                $"Response {context.Response.StatusCode} for {context.Request.Method} {context.Request.Path}"
            );
        }
    }

    public static class LoggingMiddlewareExtensions
    {
        public static IApplicationBuilder UseRabbitMQLogging(this IApplicationBuilder builder)
        {
            return builder.UseMiddleware<LoggingMiddleware>();
        }
    }
}
